### PR TITLE
util: Augment util_av to store extra data

### PR DESCRIPTION
**Pull request by a-ilango**

### CODE REVIEW

File: prov/util/src/util_av.c [hunk 1]
```diff
@@ -415,23 +426,32 @@ static int util_av_init(struct util_av *av, const struct fi_av_attr *attr,
...
 			  (av->hash.total_count * sizeof(*av->hash.table)));
-	if (!av->data)
+	if (!av->addr)
 		return -FI_ENOMEM;
 
+	av->data = calloc(av->count, sizeof *(av->data));
+	if (!av->data) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
```
Comment (shefty) [inline prov/util/src/util_av.c]: We should probably make the data allocation optional - maybe controlled through an internal flag?

**Comment by shefty** [comment]
Thinking about this more, a few things occurred to me.  One, if this proceeds, the data should probably be located in memory next to the address.  The second is that this is basically just adding a wrapper around a uint64 array, with very little abstraction.  Finally, I'm not sure if this ends up being useful if we're dealing with a shared AV.  It's likely each process will need their own private data.

It seems that there's some sort of commonality needed to support layered providers, or even a non-layered provider that needs to send internal messages (such as acks) to a sender that is not in the AV.  If we end up layering one AV over another, we're going to lose any benefit to having a shared AV exported to the app.

If we add in maintaining connection states as part of the AV, the common functionality that we need may grow even more.  I wonder if we need to re-design the base util AV to handle more cases than what was needed by just the UDP provider.

**Comment by jithinjosepkl** [comment]
+1 on storing the data next to address. Also, if it does not add too much overhead, may be we should add lookup based on data field. It will be beneficial for the layered provider (atleast for rxd) for the reverse lookup.

**Comment by shefty** [comment]
It turns out that we can't place the data next to the address, or we can't support shared AVs.  The data is unique per process.  Arun and I had offline discussions and I think we have a generlc proposal for a 'communication map' that will work for all layered providers, including sockets.  Arun is implementing it to verify that it works for him.

**Comment by shefty** [comment]
This PR is being replaced by a different architecture and new utility abstraction.

**Pull request closed without merging**